### ISSUE TITLE
Unpredictable release() failures prevent reactor-style integration

### ISSUE THREAD

**Issue by b-spencer** [chriskohlhoff/asio#1097]
On Windows, asio socket types support a `release()` method to disown the native
handle `SOCKET`.  However, even on modern Windows, the `release()` method fails
at "unpredictable" times inside `NtSetInformationFile()` with
`STATUS_UNSUCCESSFUL`, perhaps depending on the socket state.  This has
implications for driving real-world reactor-style code (such as in a third-party
library) from an asio `io_context`.

The failure of `release()` stems from the use of Windows IO Completion Ports for
`wait_read` events (but not other event types) on stream sockets.  Asio
responsibly tries to optimize monitoring for these reactor-style events with its
IOCP service where it can.  Windows IOCP seems to only support the technique
that asio uses for stream sockets, so `release()` fails when using `tcp::socket`
but not when using `udp::socket`.

In fact, if you "cheat" and use `udp::socket` to monitor a TCP `SOCKET` with
`async_wait()`, asio will avoid the IOCP implementation of `wait_read` and use
its select reactor instead.  If all you do is on this "mismatched" `udp::socket`
is call `async_wait()` or other `basic_socket` operations, you can successfully
monitor for events and use `release()` without unpredictable failures.

This works because, on Windows, datagram sockets can't implement `wait_read` in
terms of IOCP, so asio uses an internal `select()`-based reactor.
(Alternatively, you can specialize `basic_socket` to lie about the `type` of the
socket and thus fool `win_iocp_socket_service_base` into thinking TCP `SOCKET`
handles are `SOCK_DGRAM` and thus force it to use the select reactor, but the
result is the same.)

While clearly the failure of `NtSetInformationFile()` is beyond asio's control,
it does have at least one important practical implication: 

> On Windows, you cannot integrate reactor-style libraries that own their
> sockets into an asio `io_context` via `async_wait()` without detailed awareness
> of the failure modes of `release()` in real-world cases.

Without `release()` it becomes impossible for you to relinquish ownership of the
`SOCKET` from the asio socket types at the time the library expects.  Clearly,
having both asio and the other library thinking they control the lifetime of a
`SOCKET` does not end well.

Even if you are aware of the `release()` failure modes, as an integrator between
asio and the reactor-style library, you cannot necessarily do anything to handle
the case.  You can't know when each possible failure trigger has occurred
because you do not perform the reads and you do not perform the IOCP operations.

I have prepared a small program that creates one of the simple cases where
`release()` fails with `STATUS_UNSUCCESSFUL` when using IOCP.  It sets up a
situation where an `async_wait()` operation is started after the library has
performed a `recv()` that yields EOF, but where the library has _not_ _yet_
closed the socket or stopped asking the `SOCKET` to be monitored for read
events.  The program allows you to use asio's IOCP or select reactor
implementations of `wait_read` on the same type of sockets to witness the
failure or success of `release()`.

While I concede that it is "strange" for a library to not immediately and
synchronously ask for the `SOCKET` to be no longer monitored immediately after
seeing an EOF, it is _valid_ behaviour in a traditional reactor-style approach.
My guess is that Windows IOCP behaviour has a quirk where a socket that has
reported an EOF to the application but then started another asynchronous read
operation ends up in a state where `NtSetInformationFile()` can't change its
completion information.  It's not immediately clear how one can work around this
or even recover from it once it happens.

As an aside, it's worth noting as a second example that curl sometimes closes
sockets _before_ notifying the application (via its "multi" interface) that it
should stop monitoring for events, which can lead to another situation where
`release()` ceases to function.  I haven't produced a minimal test case for
this, but this kind of assumption that "the fd state doesn't matter until the
reactor runs `poll()` again" does exist in the wild.

So, it seems that the use of Windows IOCP as an implementation of `wait_read`
becomes an obstacle if you need to `release()` native handles owned elsewhere.
Unfortunately, that's exactly the circumstances you find yourself in when an
existing reactor-style library doesn't tell you about events that are immaterial
in reactor-style IO (such an EOF for one of its sockets).  So, you can easily
end up stuck with a socket you can't `release()` during normal operation,
leaving it with two owners, and disaster ensues.  Ownership that you can't
relinquish is pretty difficult to deal with.

While this seems like it might be a Windows limitation rather than a problem
with asio's implementation of IOCP, it does make reactor-style integration with
asio on Windows impractical if that reactor-style code doesn't happen to conform
to the unstated restrictions of IOCP.

Since modifying every existing third-party library's socket management logic is
impractical, the only reasonable options I've come up with are:

 - "lying" to asio in order to trick it into using the select reactor (which has
   its own limitations and inefficiencies), or
   
 - not using asio to drive reactor-style IO (which basically means multiple
   threads in some way).

Are there alternatives?

# Demo program

## `prog.cc`
```c++
//*****************************************************************************

// Set the minimum version to Windows 10 1809 to avoid "new API" warnings.  The
// exact choice doesn't matter.
#define NTTDI_VERSION NTDDI_WIN10_RS5
#include <SDKDDKVer.h>

#include <asio/io_context.hpp>
#include <asio/ip/udp.hpp>
#include <asio/ip/tcp.hpp>

#include <chrono>
#include <iostream>
#include <optional>
#include <stdexcept>
#include <string>
#include <system_error>
#include <type_traits>

using namespace std::literals;

namespace
{
  // This is C++17.
  //
  // Last tested with:
  //
  //  - asio-1.18.2 on Windows 10 20H2 using MSVC 2019 (16.6.9)
  //  - asio-1.23.0 on Windows 10 21H2 using MSVC 2022 (17.2.2)
  //
  // CL flags: -GS -GL -W3 -Gy -Zc:wchar_t -Gm- -O2 -Oi -MD -EHsc -std:c++17
  // LINK flags: -DYNAMICBASE -NXCOMPAT -OPT:REF -RELEASE -SUBSYSTEM:CONSOLE

  // See README.md.
  //
  // Set this to true to use asio::ip::tcp::socket to issue async_wait() and see
  // the failure.
  //
  // Set this to false to use udp::socket and see it work (because of the use of
  // the select reactor).
  //
  constexpr auto useIocp = true;

  // When this is true, we simulate the library reacting to the EOF by removing
  // interest in its fd _before_ a new async_wait() is scheduled on read.
  //
  // When this is false, we simulate the library leaving its read event interest
  // in place even after the EOF.
  //
  // See README.md for more details.
  //  
  constexpr auto immediateUninterestOnEof = false;
  
  // Pick the type that we will use to perform async_wait().
  using protocol = std::conditional_t<useIocp, asio::ip::tcp, asio::ip::udp>;

  // Throw if result isn't true.
  template <typename result_type>
  void check(const char* const name,
             result_type&& result)
  {
    if(!result)
    {
      throw std::runtime_error("FAIL: "s + name);
    }
    std::cout << "pass: " << name << '\n';
  }

  // A pair of connected TCP sockets.
  //
  // No matter what the "protocol" type is, we connect a pair of TCP sockets.
  // These represent "arbitrary" SOCKET handles in a reactor-mode library that
  // we are trying to integrate into asio.  So, in reality, they would not be
  // "already" in asio types, but for convenience we use asio to generate a pair
  // of connected sockets.  (The issue we are demonstrating occurs if we made
  // these sockets using the winsock API directly, without asio.)
  //
  auto connect(asio::io_context& io)
  {
    using asio::ip::tcp;
    tcp::acceptor acceptor{io, {asio::ip::address_v4::loopback(), 0}};
    std::cout << "listen at " << acceptor.local_endpoint() << '\n';

    // Connect two sockets and return them.
    tcp::socket client{io, tcp::v4()};
    client.connect(acceptor.local_endpoint());
    auto server = acceptor.accept();
    check("connected", true);
    return std::make_pair(std::move(client), std::move(server));
  }

  // Wait on socket for read with async_wait(), and then do it again, forever.
  // This represents us maintaining a reactor-style library's continuing
  // interest in read events.
  void waitRead(protocol::socket& socket)
  {
    socket.async_wait(
      protocol::socket::wait_read,
      [&socket](const auto& ec)
      {
        check("wait has no error", !ec);

        // We don't react to the read event here, but that's not the point.
        waitRead(socket);
      });
  }
}

//-----------------------------------------------------------------------------

int
main()
{
  try
  {
    // Seed ourselves with a pair of TCP sockets.  (These are always TCP sockets
    // even when we end up monitoring them with an asio::ip::udp::socket.)
    asio::io_context io;
    auto [client, server] = connect(io);

    // We're going to monitor for read events with either a TCP socket (IOCP
    // mode) or a UDP socket (select reactor).  Arrange for that to be referred
    // to as "socket" in either scenario.
    //
    // We're trying to avoid doing anything that could influence either the
    // IOCP or select reactor implementations, so we minimize moves etc., and
    // thus this is a bit clumsy.
    //
    std::optional<protocol::socket> extraSocket;
    auto& socket =
      [&io, &client, &extraSocket]() -> protocol::socket&
      {
        if constexpr(useIocp)
        {
          // We conveniently already have the TCP SOCKET handle in a
          // tcp::socket, so we don't need to do anything.
          //
          // In a real reactor-style integration, we'd be given just the raw
          // SOCKET and we'd have to "loan" it to an asio socket type for the
          // duration of the library's interest in events.  Most importantly, we
          // also need to release() it from that asio socket type when the
          // library is no longer interested in events.  The library owns the
          // SOCKET and definitely won't expect or allow us to close it.
          //
          return client;
        }
        else
        {
          // We're about to cheat and put a TCP SOCKET handle into a
          // udp::socket.  We won't do anything except async_wait() on it, so
          // it doesn't matter.  Plus, this isn't what makes it break.  This is
          // what makes it _work_.
          //
          // _This_ release() works fine (we haven't started any asynchronous
          // operations yet), and moves ownership of the SOCKET handle from
          // "client" to "extraSocket".  In a real reactor-style library
          // integration, we'd be given just the raw SOCKET and would put it in
          // an asio socket type, much like this.
          //
          extraSocket.emplace(io, protocol::v4(), SOCKET(client.release()));
          check("old client closed", !client.is_open());
          return *extraSocket;
        }
      }();
    check("socket open", socket.is_open());
    check("server open", server.is_open());

    // Simulate the library asking to be notified of reactor-style read events
    // on the file descriptor and then calling back on the library to read from
    // it.
    bool waited = false;
    socket.async_wait(
      protocol::socket::wait_read,
      [&io, &socket, &waited](const std::error_code& ec)
      {
        // This block of code simulates the library code reading from its SOCKET
        // without the help of asio, which is exactly how real reactor-style
        // libraries work.
        {
          char buf;
          const int result 
            = ::recv(socket.native_handle(), &buf, int{sizeof(buf)}, 0);

          // In our case, we expect an EOF indication.
          check("read not error", result != -1);
          check("read eof", result == 0);
        }

        // The library got an EOF on its read event.
        //
        // The normal behaviour here would be for the library to tell us that it
        // is no longer interested in any events from the socket (and then close
        // the socket), but it is not _obligated_ to do so immediately.  By
        // waiting for further read events here, we simulate the library not
        // changing its set of monitored events as an immediate, synchronous
        // reaction to the EOF (before we start another async_wait()
        // operation).  See README.md for further discussion.
        //
        // Obviously any responsible library will not _close_ the socket
        // _before_ asking for the socket to no longer be monitored.  Note that
        // we are _not_ simulating such opposite, irresponsible behaviour here.
        //
        if constexpr(!immediateUninterestOnEof)
        {
          // Our asio integration doesn't know the EOF happened, so we start a
          // new async_wait() to represent the library's continued interest in
          // read events.
          waitRead(socket);
        }

        // Remember that we got this far so we know the test setup is correct.
        waited = true;

        // And tell the loop it can stop.  (The same problem happens even if the
        // loop doesn't stop, but stopping the loop keeps this demonstration
        // program simple.)
        check("stopping loop", true);
        io.stop();
      });

    // Before we actually run the loop, close the other end of the TCP
    // connection pair, which will trigger a read event to report the EOF.
    server.close();

    // Run the loop for long enough to see this happen.  It will stop sooner.
    io.run_for(5s);

    // Confirm that we set up the test as intended.
    check("first read waited", waited);

    // This should work (on new enough versions of Windows).
    socket.release();
  }
  catch(const std::exception& ex)
  {
    // Something went wrong.
    std::cout << "FAIL: exception: " << ex.what() << '\n';
    return 1;
  }

  // All ok.
  check("finished", true);
  return 0;
}

//*****************************************************************************
```

## `demo.bat`
```bat
rem Run this in a VS command prompt.

rem Start clean.
del prog.obj prog.exe

rem Build it.
rem
rem This expects asio headers to be in the current directory as if
rem the repository was cloned here.
SET CL=/GS /GL /W3 /Gy /Zc:wchar_t /Gm- /O2 /Oi /MD /EHsc /std:c++17 /I asio\asio\include
SET LINK=/DYNAMICBASE /NXCOMPAT /OPT:REF /RELEASE /SUBSYSTEM:CONSOLE
cl.exe prog.cc /link

rem Run it and show the exit code in hex.
prog.exe
cmd /c exit %ErrorLevel%
echo 0x%=ExitCode%
```
